rfcs/0003-mutations.md

```diff
@@ -0,0 +1,317 @@
+
+### Mutations Manifest
+
+The subgraph manifest (`subgraph.yaml`) now has an extra property named `mutations` which is the mutations manifest.
+
+`subgraph.yaml`
+```yaml
+specVersion: ...
+...
+mutations:
+  repository: https://npmjs.com/package/...
+  schema:
+    file: ./mutations/schema.graphql
+  resolvers:
+    apiVersion: 0.0.1
+    kind: javascript/es5
+    file: ./mutations/index.js
+dataSources: ...
+...
+```
+
+Alternatively, the mutation manifest can be external like so:  
+`subgraph.yaml`
+```yaml
+specVersion: ...
+...
+mutations:
+  file: ./mutations/mutations.yaml
+dataSources: ...
+...
+```
+`mutations/mutations.yaml`
+```yaml
+repository: https://npmjs.com/package/...
+schema:
+  file: ./schema.graphql
+resolvers:
+  apiVersion: 0.0.1
+  kind: javascript/es5
+  file: ./index.js
+```
+
+### Mutations Schema
+
+The mutations schema defines all of the mutations in our subgraph. The mutations schema is a super-set of the subgraph's schema. For example, starting from a base subgraph schema:  
+`schema.graphql`
+```graphql
+type MyEntity @entity {
+  id: ID!
+  name: String!
+  value: BigInt!
+}
+```
+
+Developers can define mutations that reference these subgraph schema types. Additionally new `input` and `interface` types can be defined for the mutations to use:  
+`mutations/schema.graphql`
+```graphql
+input MyEntityOptions {
+  name: String!
+  value: BigInt!
+}
+
+interface NewNameSet {
+  oldName: String!
+  newName: String!
+}
+
+type Mutation {
+  createEntity(
+    options: MyEntityOptions!
+  ): MyEntity!
+
+  setEnityName(
+    entity: MyEntity!
+    name: String!
+  ): NewNameSet!
+}
+```
+
+`graph-cli` handles the combining, parsing, and validating of these two schemas. The `graph-cli` verifies that the mutations schema defines a `type Mutation`, that all of the mutations within it are defined in the resolvers module (see next section).  
+
+### Mutation Resolvers
+
+Each mutation within the schema must have a corresponding resolver function defined. Resolvers will be invoked by whatever engine executes the query. They are executed locally within the client application.
+
+Mutation resolvers of kind `javascript` take the form of a javascript module. This module is expected to have a default export that contains the following properties:
+  * resolvers - The mutation resolver functions.
+  * config - A collection of config generators.
+
+`mutations/index.js`
+```javascript
+const resolvers = {
+  Mutation: {
+    async createEntity (_, args, context) {
+      // Extract mutation arguments
+      const { name, value } = args.options
+
+      // Use config properties created by the
+      // config generator functions
+      const { ethereum, ipfs } = context.graph.config
+
+      // Fetch datasource addresses & abis
+      const { MyContract } = context.graph.datasources
+      await MyContract.abi
+      await MyContract.address
```

Comment:
This is giving me headaches right now. Can we leave out access to the data sources from the initial RFC? I need to think more about the interaction between the mutations and query engine / subgraph metadata.

Reply:
We may even be able to leave this out entirely. The mutation resolvers have to have built-in knowledge about which ABIs they interact with, and the contract addresses could be passed in as arguments.

Reply:
Sure thing, would love to know more about what's erking you about this.

IMO this is an extremely valuable feature for mutations, as managing your contract addresses is a hassle when developing protocols & dApps. Having the subgraph be the source of truth would be wonderful.